homeassistant/components/marstek/sensor.py

```diff
@@ -0,0 +1,307 @@
+            self._attr_icon = "mdi:solar-power"
+        elif metric_type == "voltage":
+            self._attr_native_unit_of_measurement = UnitOfElectricPotential.VOLT
+            self._attr_icon = "mdi:flash"
+        elif metric_type == "current":
+            self._attr_native_unit_of_measurement = UnitOfElectricCurrent.AMPERE
+            self._attr_icon = "mdi:current-ac"
+        elif metric_type == "state":
+            self._attr_icon = "mdi:state-machine"
+            self._attr_device_class = None
+            self._attr_state_class = None
+        else:
+            self._attr_icon = "mdi:solar-panel"
+
+        if metric_type != "state":
+            self._attr_state_class = SensorStateClass.MEASUREMENT
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        device_ip = self._device_info.get("ip", "Unknown")
+        metric_name = self._metric_type.replace("_", " ").title()
+        return f"PV{self._pv_channel} {metric_name} ({device_ip})"
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the PV metric value."""
+        if not self.coordinator.data:
+            return None
+        value = self.coordinator.data.get(self._sensor_type)
+        if isinstance(value, (int, float)):
+            return cast(StateType, value)
+        return None
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: MarstekConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up Marstek sensors based on a config entry."""
+    device_ip = config_entry.data["host"]
+    _LOGGER.info("Setting up Marstek sensors: %s", device_ip)
+
+    udp_client = config_entry.runtime_data
+
+    device_info = {
+        "ip": config_entry.data["host"],
+        "mac": config_entry.data.get("mac", ""),
+        "device_type": config_entry.data.get("device_type", "Unknown"),
+        "version": config_entry.data.get("version", 0),
+        "wifi_name": config_entry.data.get("wifi_name", ""),
+        "wifi_mac": config_entry.data.get("wifi_mac", ""),
+        "ble_mac": config_entry.data.get("ble_mac", ""),
+    }
+
+    coordinator = MarstekDataUpdateCoordinator(
+        hass, config_entry, udp_client, device_info["ip"]
+    )
+    await coordinator.async_config_entry_first_refresh()
+
+    sensors: list[MarstekSensor] = [
+        MarstekBatterySensor(coordinator, device_info),
+        MarstekPowerSensor(coordinator, device_info),
+        MarstekDeviceModeSensor(coordinator, device_info),
+        MarstekBatteryStatusSensor(coordinator, device_info),
+        MarstekDeviceInfoSensor(coordinator, device_info, "device_ip"),
+        MarstekDeviceInfoSensor(coordinator, device_info, "device_version"),
+    ]
+
+    sensors.extend(
+        MarstekPVSensor(coordinator, device_info, pv_channel, metric_type)
+        for pv_channel in range(1, 5)
+        for metric_type in ("power", "voltage", "current", "state")
+    )
```

Comment:
Not all devices support this right?


```suggestion
    if device_info.get("device") in [ ... ]
        sensors.extend(
            MarstekPVSensor(coordinator, device_info, pv_channel, metric_type)
            for pv_channel in range(1, 5)
            for metric_type in ("power", "voltage", "current", "state")
        )
```